base/src/staged_ledger_diff.rs

```diff
@@ -9,46 +9,53 @@
 
 use crate::numbers::Amount;
 use crate::snark_work::TransactionSnarkWork;
-use crate::types::TokenId;
+use crate::types::{SignedCommandPayload, TokenId};
 use crate::user_commands::UserCommand;
+use crate::verifiable::Verifiable;
 
-use proof_systems::mina_signer::CompressedPubKey;
+use proof_systems::mina_signer::{CompressedPubKey, Signer};
 
 #[derive(Clone, PartialEq, Debug, Default)]
 /// Top level wrapper type for a StagedLedgerDiff
 pub struct StagedLedgerDiff {
-    pub diff: StagedLedgerDiffTuple,
+    pub diff_two: StagedLedgerPreDiff,
+    pub diff_one: Option<StagedLedgerPreDiff>,
 }
 
-#[derive(Clone, PartialEq, Debug, Default)]
-pub struct StagedLedgerDiffTuple(
-    pub(crate) (StagedLedgerPreDiffTwo, Option<StagedLedgerPreDiffOne>),
-);
-
-impl StagedLedgerDiffTuple {
-    pub fn diff_two(&self) -> &StagedLedgerPreDiffTwo {
-        &self.0 .0
-    }
-
-    pub fn diff_one(&self) -> &Option<StagedLedgerPreDiffOne> {
-        &self.0 .1
+impl<CTX> Verifiable<CTX> for StagedLedgerDiff
+where
+    CTX: Signer<SignedCommandPayload>,
+{
+    // StagedLedgerDiff is considered valid if:
+    // - All PreDiffs are valid
+    fn verify(&self, ctx: &mut CTX) -> bool {
+        if let Some(diff_one) = &self.diff_one {
+            diff_one.verify(ctx) && self.diff_two.verify(ctx)
+        } else {
+            self.diff_two.verify(ctx)
+        }
     }
 }
 
 #[derive(Clone, PartialEq, Debug, Default)]
-pub struct StagedLedgerPreDiffOne {
+pub struct StagedLedgerPreDiff {
     pub completed_works: Vec<TransactionSnarkWork>,
     pub commands: Vec<UserCommandWithStatus>,
     pub coinbase: CoinBase,
     pub internal_command_balances: Vec<InternalCommandBalanceData>,
 }
 
-#[derive(Clone, PartialEq, Debug, Default)]
-pub struct StagedLedgerPreDiffTwo {
-    pub completed_works: Vec<TransactionSnarkWork>,
-    pub commands: Vec<UserCommandWithStatus>,
-    pub coinbase: CoinBase,
-    pub internal_command_balances: Vec<InternalCommandBalanceData>,
+impl<CTX> Verifiable<CTX> for StagedLedgerPreDiff
+where
+    CTX: Signer<SignedCommandPayload>,
+{
+    // PreDiff is considered valid if:
+    // - all commands are valid
+    fn verify(&self, ctx: &mut CTX) -> bool {
+        self.commands
```

Comment:
Do we only verify `commands` field for this?

Reply:
That and the snark work but I guess that has to wait

Reply:
can we add a todo in this case and mention what's blocking it